blog-post/blog-post.md

```diff
@@ -0,0 +1,522 @@
+
+```
+instance word : Sum {
+    function sum (x : word, y : word) -> word {
+        let res : word ;
+        assembly {
+            res := add(x,y);
+        }
+        return res;
+    }
+}
+```
+The type `word` correspond to `uint256` in Classic Solidity. Another
+more interesting example is the instance for `uint128`, which is
+defined as follows:
+
+```
+data uint128 = uint128(word);
+
+instance uint128 : Sum {
+    function sum(x : uint128, y : uint128) -> uint128 {
+        let res : word;
+        match x, y {
+        | uint128(n), uint128(m) =>
+            assembly {
+                res := add(n,m);
+                if lt(res, n) {
+                    revert(0,0);
+                }
+                if gt(res, 0xffffffffffffffffffffffffffffffff) {
+                    revert(0,0);
+                }
+            }
+        }
+        return uint128(res);
+    }
+}
+```
+
+We start by defining a new type for `uint128` which will, internally,
+hold a `word` value. The implementation of `sum` for `uint128`
+uses pattern matching to extract the internal `word` values and
+performs the addition in a Yul block, which reverts if the result
+is not a valid unsigned 128 bits integer.
+
+Instances can also be defined over polymorphic types. As an example,
+consider the following implementation of `sum` for polymorphic pairs.
+
+```
+forall T1 T2 . T1 : Sum, T2 : Sum => instance (T1,T2) : Sum {
+  function sum (p1 : (T1,T2), p2 : (T1,T2)) -> (T1, T2) {
+      match p1, p2 {
+      | (x1,y1), (x2, y2) =>
+          return (Sum.sum(x1,x2), Sum.sum(y1,y2));
+    }
+  }
+}
+```
+The previous definition shows that, whenever we have instances of
+class `Sum` for types `T1` and `T2`, then we can also use function
+`Sum.sum` on pairs of such types.
+The reader should note that the last two examples use **pattern matching**
+to extract components of user defined algebraic types, which are
+another feature that will be part of Core Solidity.
+
+### Algebraic data types and pattern matching
+
+Algebraic Data Types provide a way of data modeling using sum types
+(disjoint unions) and product types (structural records). This enables the construction of
+precise types where invalid states are **unrepresentable** by design, thereby enhancing program
+correctness through the type system itself. As an example, consider the following type
+definition for an auction state.
+
+```
+data AuctionState =
+    NotStarted (word)
+  | Active (word, address)
+  | Ended (word, address)
+  | Cancelled;
+
+```
+The type has three constructors: `NotStarted` specifies that the action have not started yet and
+stores its reserved price, `Active` denotes that the auction has began and it stores the
+current highest bid and the address that made such bid, `Ended` represents that the auction has
+finished with success and it holds the highest bid and the winner address and constructor
+`Cancelled` is used when the auction has been cancelled.
```

Comment:
```suggestion
The type has three constructors: `NotStarted` specifies that the auction has not started yet and
stores its reserved price, `Active` denotes that the auction has begun and it stores the
current highest bid and the address that made such bid, `Ended` represents that the auction has
finished with success and it holds the highest bid and the winner address and constructor,
`Cancelled` is used when the auction has been cancelled.
```

Reply:
There is also one more mistyped `auction` below:
- `update the action state` -> `update the auction state`